File: quant1x/datasets/kline.rs
use crate::cache::{self, DataAdapter, Kind};
use crate::level1::KLineType;
use crate::Timestamp;
use serde::{Deserialize, Serialize};
use std::sync::Arc;

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct KLine {
    #[serde(rename = "Date")]
    pub date: String,
    #[serde(rename = "Open")]
    pub open: f64,
    #[serde(rename = "Close")]
    pub close: f64,
    #[serde(rename = "High")]
    pub high: f64,
    #[serde(rename = "Low")]
    pub low: f64,
    #[serde(rename = "Volume")]
    pub volume: f64,
    #[serde(rename = "Amount")]
    pub amount: f64,
    #[serde(rename = "Up")]
    pub up: i32,
    #[serde(rename = "Down")]
    pub down: i32,
    #[serde(rename = "Datetime")]
    pub datetime: String,
    #[serde(rename = "AdjustmentCount")]
    pub adjustment_count: i32,
}

impl KLine {
    pub fn headers() -> Vec<String> {
        vec![
            "Date".into(),
            "Open".into(),
            "Close".into(),
            "High".into(),
            "Low".into(),
            "Volume".into(),
            "Amount".into(),
            "Up".into(),
            "Down".into(),
            "Datetime".into(),
            "AdjustmentCount".into(),
        ]
    }
}

#[derive(Debug)]
pub struct DataKLine;

impl cache::Schema for DataKLine {
    fn kind(&self) -> Kind {
        crate::datasets::BaseKLine
    }
    fn owner(&self) -> String {
        crate::cache::DEFAULT_DATA_PROVIDER.to_string()
    }
    fn key(&self) -> String {
        "day".to_string()
    }
    fn name(&self) -> String {
        "日K线".to_string()
    }
    fn usage(&self) -> String {
        "日K线".to_string()
    }
}

impl DataAdapter for DataKLine {
    fn print(&self, _code: &str, _dates: &[Timestamp]) {}

    fn update(&self, code: &str, _date: Timestamp) {
        // 尝试从 level1 获取日线数据；若无法获取则降级为本地缓存（仅写入表头）的行为。
        // 构建与 C++ 等效的 kline 文件名。
        // 使用集中式的 config helper 构建 kline 缓存文件路径（与 C++ 保持一致）。
        let filename = crate::config::get_kline_filename(code, true);
        if filename.is_empty() {
            log::error!("[DataKLine] cannot build kline filename for {}", code);
            return;
        }
        log::debug!("[DataKLine] cache filename: {}", filename);

        // 确保父目录存在
        let path = std::path::Path::new(&filename);
        if let Some(parent) = path.parent() {
            if let Err(e) = std::fs::create_dir_all(parent) {
                log::error!(
                    "[DataKLine] failed to create parent dir {:?}: {}",
                    parent,
                    e
                );
                return;
            }
        }
        // 遵循 C++ 的逻辑：分页步长（security_bars_max）、回溯天数、合并及除权预处理
        const MAX_KLINE_LOOKBACK_DAYS: usize = 1;
        const SECURITY_BARS_MAX: usize = 800;

        // 先加载已有缓存（若存在），以确定起始日期和调整次数（与 C++ 保持一致）
        let cache_filename = filename.clone();
        let cache_klines: Vec<KLine> = read_kline_from_csv(&cache_filename);
        let klines_length = cache_klines.len();
        let mut klines_offset_days = MAX_KLINE_LOOKBACK_DAYS;
        let mut adjust_times = 0i32;
        // 默认起始日期：1990-12-19（市场首次上市日期）
        let mut current_start_date =
            crate::Timestamp::pre_market_time(1990, 12, 19).unwrap_or(crate::Timestamp::zero());
        if klines_length > 0 {
            if klines_offset_days > klines_length {
                klines_offset_days = klines_length;
            }
            let kline = &cache_klines[klines_length - klines_offset_days];
            // parse date back to Timestamp
            if let Ok(ts) = crate::Timestamp::parse(&kline.date) {
                current_start_date = ts;
            }
            adjust_times = kline.adjustment_count;
        }

        // 从当前时间确定结束日期（尽可能使用今日的盘前时间作为结束日期）
        let current_end_date =
            crate::Timestamp::pre_market_time_from_current(&crate::Timestamp::now())
                .unwrap_or(crate::Timestamp::now());
        // 构建日期范围
        let ts_range = crate::exchange::date_range(current_start_date, current_end_date, false);
        if ts_range.is_empty() {
            log::debug!("[DataKLine] empty date range for {}", code);
            return;
        }

        let total = ts_range.len();

        // 按日期范围分页从 level1 拉取数据（每页作为一个向量，保持页内顺序，之后再翻转页序以匹配 C++）
        let mut hs: Vec<Vec<crate::level1::SecurityBar>> = Vec::new();
        let step = SECURITY_BARS_MAX;
        let mut start_idx: usize = 0;
        while start_idx < total {
            let remaining = total - start_idx;
            let count = std::cmp::min(step, remaining) as u16;
            // 注意：start 使用日期索引，datasets 层直接调度 SecurityBars 请求
            match crate::datasets::kline_raw::fetch_kline(
                code,
                start_idx as u32,
                count,
                KLineType::RiK,
            ) {
                Some(resp) => {
                    let response_count = resp.count as usize;
                    if resp.list.is_empty() {
                        break;
                    }
                    hs.push(resp.list);
                    if response_count < count as usize {
                        break;
                    }
                    start_idx = start_idx.saturating_add(count as usize);
                }
                None => {
                    log::warn!(
                        "[DataKLine] fetch_kline returned None for {} start={}",
                        code,
                        start_idx
                    );
                    break;
                }
            }
        }

        // // 如果没有获取到任何页，则不保存CSV（空列表不保存，包括表头）
        // if hs.is_empty() {
        //     return;
        // }
        // C++ 会将分页结果反转为时间升序；保留每页内的顺序，只 reverse 外层 pages
        hs.reverse();

        // 根据抓取到的 pages 并按日期范围筛选，构建增量 K 线列表（按页内顺序迭代）
        let mut incremental_klines: Vec<KLine> = Vec::new();
        for page in hs.iter() {
            for row in page.iter() {
                // 为 bar 日期构造盘前时间戳
                let date_time =
                    crate::Timestamp::pre_market_time(row.year, row.month as u32, row.day as u32)
                        .unwrap_or(crate::Timestamp::now());
                if date_time < ts_range[0] || date_time > ts_range[total - 1] {
                    continue;
                }
                let kx = KLine {
                    date: date_time.only_date(),
                    open: row.open,
                    close: row.close,
                    high: row.high,
                    low: row.low,
                    volume: row.vol * 100.0,
                    amount: row.amount,
                    up: row.up_count as i32,
                    down: row.down_count as i32,
                    datetime: row.datetime.clone(),
                    adjustment_count: 0,
                };
                incremental_klines.push(kx);
            }
        }

        // 判断是否仅针对最新抓取的那一部分需要做预除权调整
        let is_fresh_fetch_require_adjustment = adjust_times == 1;
        // 仅从本地缓存加载除权除息数据（与 C++ 行为一致）
        let dividends = crate::datasets::xdxr::load_xdxr(code);
        if is_fresh_fetch_require_adjustment {
            apply_forward_adjustment_for_event!(&mut incremental_klines, ts_range[0], &dividends);
        }

        // 按照 C++ 的合并逻辑合并缓存与增量数据
        let mut klines: Vec<KLine> = Vec::new();
        if klines_length > klines_offset_days {
            klines.extend_from_slice(&cache_klines[..(klines_length - klines_offset_days)]);
        }
        if klines.is_empty() {
            klines = incremental_klines.clone();
        } else {
            klines.extend(incremental_klines.into_iter());
        }

        if !is_fresh_fetch_require_adjustment {
            apply_forward_adjustment_for_event!(&mut klines, ts_range[0], &dividends);
        }

        // 持久化保存
        if klines.is_empty() {
            // 空列表不保存CSV，包括表头
            return;
        }
        let tmp = format!("{}.tmp", filename);
        match std::fs::File::create(&tmp) {
            Ok(f) => {
                let mut w = csv::Writer::from_writer(f);
                if let Err(e) = w.write_record(KLine::headers()) {
                    log::error!("[DataKLine] write header failed: {}", e);
                }
                for row in klines.iter() {
                    let rec: Vec<String> = vec![
                        row.date.clone(),
                        row.open.to_string(),
                        row.close.to_string(),
                        row.high.to_string(),
                        row.low.to_string(),
                        row.volume.to_string(),
                        row.amount.to_string(),
                        row.up.to_string(),
                        row.down.to_string(),
                        row.datetime.clone(),
                        row.adjustment_count.to_string(),
                    ];
                    if let Err(e) = w.write_record(rec) {
                        log::error!("[DataKLine] write row failed: {}", e);
                    }
                }
                let _ = w.flush();
                if let Err(e) = std::fs::rename(&tmp, &filename) {
                    log::error!("[DataKLine] rename failed {} -> {}: {}", tmp, filename, e);
                }
            }
            Err(e) => {
                log::error!("[DataKLine] create tmp {} failed: {}", tmp, e);
            }
        }
    }
}

// 辅助函数：从 CSV 读取 K 线，功能类似 C++ 的 read_kline_from_csv
pub fn load_klines(filename: &str) -> Vec<KLine> {
    // 使用基于 Serde 的 CSV 反序列化，使字段名可自动与表头匹配。
    let mut klines: Vec<KLine> = Vec::new();
    match std::fs::File::open(filename) {
        Ok(f) => {
            let mut rdr = csv::ReaderBuilder::new().has_headers(true).from_reader(f);
            // Deserialize all records into Vec<KLine> in one go; CSV+Serde maps headers -> struct fields
            match rdr
                .deserialize::<KLine>()
                .collect::<Result<Vec<KLine>, csv::Error>>()
            {
                Ok(v) => klines = v,
                Err(e) => log::error!(
                    "[DataKLine] failed to deserialize kline file {}: {}",
                    filename,
                    e
                ),
            }
        }
        Err(_) => { /* missing file -> return empty vector */ }
    }
    klines
}

fn read_kline_from_csv(filename: &str) -> Vec<KLine> {
    load_klines(filename)
}

pub fn init() {
    let plugin = Arc::new(DataKLine) as Arc<dyn DataAdapter>;
    cache::register(plugin);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_kline_update() {
        let adapter = DataKLine;
        let code = "sz002350";
        let date = Timestamp::now();
        adapter.update(code, date);
    }
}

File: quant1x/datasets/kline.go
// 单一实现（基于 C++ datasets/kline.cpp）
package datasets

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"path/filepath"
	"strconv"
	"strings"

	"gitee.com/quant1x/quant1x/quant1x/config"
	"gitee.com/quant1x/quant1x/quant1x/data"
	"gitee.com/quant1x/quant1x/quant1x/exchange"
	"gitee.com/quant1x/quant1x/quant1x/level1"
)

// KLine 对应 C++ 中的 datasets::KLine，用于表示日线数据。
type KLine struct {
	Date            string  `csv:"date"`             // 日期 YYYY-MM-DD
	Open            float64 `csv:"open"`             // 开盘价
	Close           float64 `csv:"close"`            // 收盘价
	High            float64 `csv:"high"`             // 最高价
	Low             float64 `csv:"low"`              // 最低价
	Volume          float64 `csv:"volume"`           // 成交量(股)
	Amount          float64 `csv:"amount"`           // 成交额(元)
	Up              int     `csv:"up"`               // 涨家数
	Down            int     `csv:"down"`             // 跌家数
	Datetime        string  `csv:"datetime"`         // 日期时间 YYYY-MM-DD HH:MM:SS
	AdjustmentCount int     `csv:"adjustment_count"` // 复权次数
}

const maxKlineLookbackDays = 1

// FetchKLines 从 level1 拉取 K 线数据并转换为 datasets.KLine。
func FetchKLines(securityCode string, category level1.KLineType, start, count uint16) ([]KLine, error) {
	conn, release, err := level1.GetStdConnection()
	if err != nil {
		return nil, fmt.Errorf("level1 client acquire failed: %w", err)
	}
	if release != nil {
		defer release()
	}
	if conn == nil || conn.Conn() == nil {
		return nil, fmt.Errorf("nil connection from level1 client")
	}

	req := level1.NewSecurityBarsRequest(securityCode, category, start, count)
	resp := level1.NewSecurityBarsResponse(req.IsIndex, uint16(req.Param.Category))

	if err := level1.Process(conn, req, resp); err != nil {
		return nil, fmt.Errorf("security bars request failed: %w", err)
	}

	out := make([]KLine, 0, len(resp.List))
	for _, b := range resp.List {
		date := b.DateTime
		if idx := strings.Index(b.DateTime, " "); idx >= 0 {
			date = b.DateTime[:idx]
		}
		kl := KLine{
			Date:  date,
			Open:  b.Open,
			Close: b.Close,
			High:  b.High,
			Low:   b.Low,
			// level1 返回的成交量单位为"手"，与 C++ 保持一致：卷转换为股时使用 vol * 100
			Volume:          b.Vol * 100,
			Amount:          b.Amount,
			Up:              int(b.UpCount),
			Down:            int(b.DownCount),
			Datetime:        b.DateTime,
			AdjustmentCount: 0,
		}
		out = append(out, kl)
	}
	return out, nil
}

// CumulativeAdjustment 表示累计复权因子。
// 字段与 C++ 中的 `factors::CumulativeAdjustment` 保持一致。
type CumulativeAdjustment struct {
	M                    float64 // multiplicative factor
	A                    float64 // additive factor
	ShareAdjustmentRatio float64 // share adjustment ratio for volumes
	No                   int     // number of adjustments applied
}

// Adjust 根据复权因子对 KLine 执行前复权调整。
func (k *KLine) Adjust(adj CumulativeAdjustment) {
	// compute adjusted prices
	k.Open = k.Open*adj.M + adj.A
	k.Close = k.Close*adj.M + adj.A
	k.High = k.High*adj.M + adj.A
	k.Low = k.Low*adj.M + adj.A

	// compute average price before changing volume (matches C++ logic)
	var ap float64
	if k.Volume != 0 {
		ap = k.Amount / k.Volume
	}
	apAdjusted := ap*adj.M + adj.A

	// adjust volume
	k.Volume = k.Volume * (1.0 + adj.ShareAdjustmentRatio)

	// recalc amount using adjusted avg price and new volume
	k.Amount = k.Volume * apAdjusted

	k.AdjustmentCount = adj.No
}

// SaveKline 将 K 线写入 CSV 文件（包含表头）。
func SaveKline(filename string, values []KLine) error {
	f, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer f.Close()
	w := csv.NewWriter(f)
	defer w.Flush()

	header := []string{"date", "open", "close", "high", "low", "volume", "amount", "up", "down", "datetime", "adjustment_count"}
	if err := w.Write(header); err != nil {
		return err
	}
	for _, r := range values {
		row := []string{
			r.Date,
			fmt.Sprintf("%g", r.Open),
			fmt.Sprintf("%g", r.Close),
			fmt.Sprintf("%g", r.High),
			fmt.Sprintf("%g", r.Low),
			fmt.Sprintf("%g", r.Volume),
			fmt.Sprintf("%g", r.Amount),
			fmt.Sprintf("%d", r.Up),
			fmt.Sprintf("%d", r.Down),
			r.Datetime,
			fmt.Sprintf("%d", r.AdjustmentCount),
		}
		if err := w.Write(row); err != nil {
			return err
		}
	}
	return nil
}

// ReadKlineFromCSV 从 CSV 文件中读取 KLine 条目；解析错误会返回错误。
func ReadKlineFromCSV(filename string) ([]KLine, error) {
	f, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	r := csv.NewReader(f)
	rows, err := r.ReadAll()
	if err != nil {
		return nil, err
	}
	if len(rows) == 0 {
		return nil, nil
	}
	// 期望第一行为表头
	out := make([]KLine, 0, len(rows)-1)
	for i := 1; i < len(rows); i++ {
		rec := rows[i]
		if len(rec) < 11 {
			continue
		}
		open, _ := strconv.ParseFloat(rec[1], 64)
		closev, _ := strconv.ParseFloat(rec[2], 64)
		high, _ := strconv.ParseFloat(rec[3], 64)
		low, _ := strconv.ParseFloat(rec[4], 64)
		vol, _ := strconv.ParseFloat(rec[5], 64)
		amt, _ := strconv.ParseFloat(rec[6], 64)
		up, _ := strconv.Atoi(rec[7])
		down, _ := strconv.Atoi(rec[8])
		datetime := rec[9]
		adjCount, _ := strconv.Atoi(rec[10])
		k := KLine{
			Date:            rec[0],
			Open:            open,
			Close:           closev,
			High:            high,
			Low:             low,
			Volume:          vol,
			Amount:          amt,
			Up:              up,
			Down:            down,
			Datetime:        datetime,
			AdjustmentCount: adjCount,
		}
		out = append(out, k)
	}
	return out, nil
}

// DataKLine 实现了与 C++ DataKLine 类似的缓存适配器更新逻辑
type DataKLine struct{}

// 实现 data.DataAdapter 的 Schema 方法
func (d *DataKLine) Kind() data.Kind { return BaseKLine }
func (d *DataKLine) Owner() string   { return data.DefaultDataProvider }
func (d *DataKLine) Key() string     { return "kline" }
func (d *DataKLine) Name() string    { return "日K线" }
func (d *DataKLine) Usage() string   { return "" }

// Print 实现 data.DataAdapter.Print（可变参数日期）
func (d *DataKLine) Print(code string, dates ...exchange.Timestamp) {
	_ = code
	_ = dates
}

// fetchRawSecurityBars 执行底层 level1 SecurityBars 请求并返回原始响应列表（未转换）。
func fetchRawSecurityBars(securityCode string, category level1.KLineType, start, count uint16) ([]level1.SecurityBar, error) {
	conn, release, err := level1.GetStdConnection()
	if err != nil {
		return nil, fmt.Errorf("level1 client acquire failed: %w", err)
	}
	if release != nil {
		defer release()
	}
	if conn == nil || conn.Conn() == nil {
		return nil, fmt.Errorf("nil connection from level1 client")
	}

	req := level1.NewSecurityBarsRequest(securityCode, category, start, count)
	resp := level1.NewSecurityBarsResponse(req.IsIndex, uint16(req.Param.Category))
	if err := level1.Process(conn, req, resp); err != nil {
		return nil, fmt.Errorf("security bars request failed: %w", err)
	}
	return resp.List, nil
}

// Update 对应 C++ DataKLine::Update 的行为：读取本地缓存、确定时间窗口、分页拉取 level1 数据、
// 反转与合并结果、在适当时机应用前复权，并写回缓存文件。
func (d *DataKLine) Update(code string, _date exchange.Timestamp) {
	// 1. 确定缓存文件并读取本地缓存
	cacheFilename := config.GetKlineFilename(code, true)
	cacheKLines, _ := ReadKlineFromCSV(cacheFilename) // ignore parse errors, follow C++ behavior
	klinesLength := len(cacheKLines)
	klinesOffsetDays := maxKlineLookbackDays
	adjustTimes := 0

	// 默认起始日期（使用 datasets.MarketFirstDate，与 C++ 的 market_first_date 等价）
	currentStartDate := MarketFirstDate
	if klinesLength > 0 {
		if klinesOffsetDays > klinesLength {
			klinesOffsetDays = klinesLength
		}
		// use the cached kline at offset as the start date
		idx := klinesLength - klinesOffsetDays
		if idx >= 0 && idx < klinesLength {
			if ts, err := exchange.ParseTimestamp(cacheKLines[idx].Date); err == nil {
				currentStartDate = ts.PreMarketTime()
			}
		}
		adjustTimes = cacheKLines[klinesLength-1].AdjustmentCount
	}

	// 2. 确定结束日期
	// 使用当前时间的盘前时间作为结束日期，并生成每日序列作为拉取区间（C++ 使用交易日历的 date_range）
	currentEndDate := exchange.NowTimestamp().PreMarketTime()
	startT := currentStartDate
	endT := currentEndDate
	var ts []exchange.Timestamp
	for t := startT; !t.Greater(endT); t = t.Offset(24, 0, 0, 0) {
		ts = append(ts, t)
	}
	total := len(ts)
	if total == 0 {
		log.Printf("[DataKLine] empty date range for %s", code)
		return
	}

	// 3. 分页从 level1 拉取数据
	step := uint16(level1.SecurityBarsMax)
	var hs [][]level1.SecurityBar
	var elementCount int
	var start uint16 = 0
	for {
		var count uint16 = step
		if uint16(total)-start >= step {
			count = step
		} else {
			count = uint16(uint16(total) - start)
		}
		if count == 0 {
			break
		}
		reply, err := fetchRawSecurityBars(code, level1.KLineDaily, start, count)
		if err != nil {
			log.Printf("[DataKLine] fetch error for %s start=%d count=%d: %v", code, start, count, err)
			break
		}
		if len(reply) == 0 {
			break
		}
		elementCount += len(reply)
		hs = append(hs, reply)
		if len(reply) < int(count) {
			break
		}
		start += count
		if int(start) >= total {
			break
		}
	}

	// 4. 反转分段数据（服务器按最新->最旧返回）
	for i, j := 0, len(hs)-1; i < j; i, j = i+1, j-1 {
		hs[i], hs[j] = hs[j], hs[i]
	}

	// 5. 转换为增量 K 线并调整单位（成交量单位从手转为股）
	incremental := make([]KLine, 0, elementCount)
	for _, vec := range hs {
		for _, row := range vec {
			dts := exchange.PreMarketTimestamp(row.Year, row.Month, row.Day)
			if dts.Less(startT) || dts.Greater(endT) {
				continue
			}
			kx := KLine{
				Date:            dts.OnlyDate(),
				Open:            row.Open,
				Close:           row.Close,
				High:            row.High,
				Low:             row.Low,
				Volume:          row.Vol * 100,
				Amount:          row.Amount,
				Up:              int(row.UpCount),
				Down:            int(row.DownCount),
				Datetime:        row.DateTime,
				AdjustmentCount: 0,
			}
			incremental = append(incremental, kx)
		}
	}

	// 6. 判断是否需要复权处理
	isFreshFetchRequireAdjustment := adjustTimes == 1
	dividends, _ := LoadXdxr(code)
	if isFreshFetchRequireAdjustment {
		ApplyForwardAdjustmentForEvent(incremental, currentStartDate.OnlyDate(), dividends)
	}

	// 7. 合并本地缓存与增量数据
	var klines []KLine
	if klinesLength > klinesOffsetDays {
		klines = append(klines, cacheKLines[:klinesLength-klinesOffsetDays]...)
	}
	if len(klines) == 0 {
		klines = incremental
	} else {
		klines = append(klines, incremental...)
	}

	// 8. 若不是仅更新最新记录，则对全量数据做前复权处理
	if !isFreshFetchRequireAdjustment {
		ApplyForwardAdjustmentForEvent(klines, currentStartDate.OnlyDate(), dividends)
	}

	// 9. 保存缓存（确保父目录存在）
	if err := os.MkdirAll(filepath.Dir(cacheFilename), 0o755); err != nil {
		log.Printf("[DataKLine] failed to create parent dir for %s: %v", cacheFilename, err)
		return
	}
	if err := SaveKline(cacheFilename, klines); err != nil {
		log.Printf("[DataKLine] save_kline failed: %v", err)
	}
}

func init() {
	// register DataKLine plugin (ignore error if already registered)
	_ = data.Register(&DataKLine{})
}
